src/main/kotlin/cacophony/grammars/Grammar.kt

```diff
@@ -1,13 +1,27 @@
 package cacophony.grammars
 
+import cacophony.automata.joinAutomata
+import cacophony.automata.minimalization.buildDFAFromRegex
 import cacophony.utils.AlgebraicRegex
 
 data class Production<SymbolType>(
     val lhs: SymbolType,
     val rhs: AlgebraicRegex<SymbolType>,
 )
 
+infix fun <SymbolType> SymbolType.produces(regex: AlgebraicRegex<SymbolType>) = Production(this, regex)
+
 data class Grammar<SymbolType>(
     val start: SymbolType,
     val productions: Collection<Production<SymbolType>>,
-)
+) {
+    fun buildAutomata() = productions.groupBy { it.lhs }.mapValues { buildAutomaton(it.value) }
```

Comment:
im not sure why you want this? Grammar should be just a bare grammar definition, as for things we build from this definition we have `AnalyzedGrammar`